### PR TITLE
Add Mapped Condition

### PR DESCRIPTION
**Pull request by joel-costigliola**
Fixes #2103

### CODE REVIEW

**Comment by joel-costigliola** [comment]
@stbischof I have worked on your PR and created this one, can you review it (your turn! :smile:) to check if I have properly captured your idea - please check the javadoc.

A few notable points:
- I have renamed `mapped` to `mappedCondition` which feels more expressive to me.
- I have revamped the javadoc
- refactor the tests.

**Comment by stbischof** [comment]
I will have a look. But at first I would say thank you very much!

**Review by stbischof — CHANGES_REQUESTED** [review]
just found  the old name in the JavaDoc

File: src/main/java/org/assertj/core/condition/MappedCondition.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+   * @param <TO> the type of object the nested condition accept.
+   * @param mapping the Function that maps the value to test to the a value for the nested condition.
+   * @param mappingDescription describes the mapping verbal.
+   * @param condition the nested condition to evaluate.
+   * @return the created {@code Mapped}.
+   * @throws NullPointerException if the given condition is {@code null}.
+   * @throws NullPointerException if the given mapping is {@code null}.
+   */
+  public static <FROM, TO> MappedCondition<FROM, TO> mappedCondition(Function<FROM, TO> mapping, String mappingDescription,
+                                                                     Condition<TO> condition) {
```
Comment (nach-o-man) [inline src/main/java/org/assertj/core/condition/MappedCondition.java:63]: @joel-costigliola just curious if there is any reason **mappingDescription** is not last argument? That way a varargs method can be introduced like in Condition class.

Comment (joel-costigliola) [reply src/main/java/org/assertj/core/condition/MappedCondition.java:63]: good idea!

**Comment by joel-costigliola** [comment]
@stbischof @nach-o-man I have further simplified the code by having only one method to set the mapping description and made it protected so that it can be overridden.
I also simplified a bit the description itself, check the latest commit https://github.com/assertj/assertj-core/pull/2193/commits/0a0af1a587a5767fa1188c63ca80d6f5ec0a2e96

**Comment by stbischof** [comment]
Lgtm

**Comment by nach-o-man** [comment]
> Lgtm

I sevond that.

**Comment by joel-costigliola** [comment]
thanks for the review!

**Pull request merged**